File: methods/matlab/logistic_regression.py
'''
  @file logistic_regression.py
  @author Marcus Edel

  Class to benchmark the matlab Logistic Regression method.
'''

import os
import sys
import inspect

# Import the util path, this method even works if the path contains symlinks to
# modules.
cmd_subfolder = os.path.realpath(os.path.abspath(os.path.join(
  os.path.split(inspect.getfile(inspect.currentframe()))[0], "../../util")))
if cmd_subfolder not in sys.path:
  sys.path.insert(0, cmd_subfolder)

#Import the metrics definitions path.
metrics_folder = os.path.realpath(os.path.abspath(os.path.join(
  os.path.split(inspect.getfile(inspect.currentframe()))[0], "../metrics")))
if metrics_folder not in sys.path:
  sys.path.insert(0, metrics_folder)

from log import *
from profiler import *
from definitions import *

import shlex
import subprocess
import re
import collections

'''
This class implements the Logistic Regression benchmark.
'''
class LogisticRegression(object):

  '''
  Create the Logistic Regression benchmark instance.

  @param dataset - Input dataset to perform Logistic Regression on.
  @param timeout - The time until the timeout. Default no timeout.
  @param path - Path to the matlab binary.
  @param verbose - Display informational messages.
  '''
  def __init__(self, dataset, timeout=0, path=os.environ["MATLAB_BIN"],
      verbose=True):
    self.verbose = verbose
    self.dataset = dataset
    self.path = path
    self.timeout = timeout

  '''
  Destructor to clean up at the end. Use this method to remove created files.
  '''
  def __del__(self):
    Log.Info("Clean up.", self.verbose)
    filelist = ["predictions.csv", "matlab_lr_probs.csv"]
    for f in filelist:
      if os.path.isfile(f):
        os.remove(f)

  '''
  Logistic Regression benchmark instance. If the method has been successfully
  completed return the elapsed time in seconds.

  @param options - Extra options for the method.
  @return - Elapsed time in seconds or a negative value if the method was not
  successful.
  '''
  def RunMetrics(self, options):
    Log.Info("Perform Logistic Regression.", self.verbose)

    # No options accepted for this script.
    if len(options) > 0:
      Log.Fatal("Unknown parameters: " + str(options))
      raise Exception("unknown parameters")

    # If the dataset contains two files then the second file is the test
    # file. In this case we add this to the command line.
    if len(self.dataset) == 2:
      inputCmd = "-i " + self.dataset[0] + " -t " + self.dataset[1]
    else:
      inputCmd = "-i " + self.dataset[0]

    # Split the command using shell-like syntax.
    cmd = shlex.split(self.path + "matlab -nodisplay -nosplash -r \"try, " +
        "LOGISTIC_REGRESSION('"  + inputCmd + "'), catch, exit(1), end, exit(0)\"")

    # Run command with the nessecary arguments and return its output as a byte
    # string. We have untrusted input so we disable all shell based features.
    try:
      s = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=False,
          timeout=self.timeout)
    except subprocess.TimeoutExpired as e:
      Log.Warn(str(e))
      return -2
    except Exception as e:
      Log.Fatal("Could not execute command: " + str(cmd))
      return -1

    # Datastructure to store the results.
    metrics = {}

    # Parse data: runtime.
    timer = self.parseTimer(s)

    if timer != -1:
      metrics['Runtime'] = timer.total_time

      Log.Info(("total time: %fs" % (metrics['Runtime'])), self.verbose)

    return metrics

  '''
  Parse the timer data form a given string.

  @param data - String to parse timer data from.
  @return - Namedtuple that contains the timer data or -1 in case of an error.
  '''
  def parseTimer(self, data):
    # Compile the regular expression pattern into a regular expression object to
    # parse the timer data.
    pattern = re.compile(br"""
        .*?total_time: (?P<total_time>.*?)s.*?
        """, re.VERBOSE|re.MULTILINE|re.DOTALL)

    match = pattern.match(data)
    if not match:
      Log.Fatal("Can't parse the data: wrong format")
      return -1
    else:
      # Create a namedtuple and return the timer data.
      timer = collections.namedtuple("timer", ["total_time"])

      return timer(float(match.group("total_time")))

File: methods/matlab/LOGISTIC_REGRESSION.m
% @file LOGISTIC_REGRESSION.m
% @author Marcus Edel
%
% Logistic Regression with matlab.

function logistic_regression(cmd)
% Logistic Regression and Prediction.
%
% Required options:
%     (-i) [string]    File containing X (predictors).
%
% Options:
%     (-r) [string]    File containing y (responses). If not given, the
%                      responses are assumed to be the last row of the
%                      input file.
%     (-t) [string]    File containing test dataset.
%     (-o) [string]    This file is where the predicted responses will be
%                      saved

% Load input dataset.
regressorsFile = regexp(cmd, '.*?-i ([^\s]+)', 'tokens', 'once');
responsesFile = regexp(cmd, '.*?-r ([^\s]+)', 'tokens', 'once');
testFile = regexp(cmd, '.*?-t ([^\s]+)', 'tokens', 'once');
estimatesFile = regexp(cmd, '.*?-o ([^\s]+)', 'tokens', 'once');

X = csvread(regressorsFile{:});

if isempty(responsesFile)
  y = X(:,end) + 1; % We have to increment because labels must be positive.
  X = X(:,1:end-1);
else
  y = csvread(responsesFile{:});
end

% Perform logistic regression.
total_time = tic;
B = mnrfit(X, y);

if ~isempty(testFile)
    % Predicted the classes.
    testSet = csvread(testFile{:});
    predictions = mnrval(B, testSet);
    % Map the probabilities to the actual classes.
    [~, idx] = max(predictions, [], 2);
end

disp(sprintf('[INFO ]   total_time: %fs', toc(total_time)))

if ~isempty(testFile)
    csvwrite('predictions.csv', idx - 1); % Subtract extra label bit.
    csvwrite('matlab_lr_probs.csv', predictions);
end

if ~isempty(estimatesFile)
    csvwrite(estimatesFile{:});
end

end
